src/puppetlabs/puppetdb/jdbc.clj

```diff
@@ -67,20 +67,112 @@
   ([table columns values]
    (sql/insert! *db* table columns values {})))
 
+(defn- insert-multi-row
+  "Given a table and a list of columns, followed by a list of column
+  value sequences, return a vector of the SQL needed for the insert
+  followed by the list of column value sequences. The entities
+  function specifies how column names are transformed."
+  [table columns values {:keys [on-conflict] :as opts}]
+  (let [nc (count columns)
+        vcs (map count values)]
+    (if (not (and (or (zero? nc) (= nc (first vcs))) (apply = vcs)))
```

Comment:
Am I reading this right?
If nc == 0 and vcs are equal, then this won't throw an IllegalArgException? Is that intended?

This might be easier to read as (if I'm understanding the intention)
```
(if (apply = (conj vcs nc)) ...) 
```
I'm not really sure what the (zero? nc) is trying to capture though

Reply:
The new jdbc functions in here are ripped off from https://github.com/clojure/java.jdbc/blob/master/src/main/clojure/clojure/java/jdbc.clj#L1392 . Basically the private functions needed to be modified for on-conflict support. I'll add comments to make that explicit.

Assuming there's no bugs in them I think it's probably better to let them be.

Reply:
there might be a larger question as to whether we want to introduce on-conflict with this work. It's only used in storage/realize-paths, and it would be feasible to do two roundtrips like we previously have.